colossalai/nn/optimizer/distributed_came.py

```diff
@@ -0,0 +1,284 @@
+            final gradient update (default: 1.0)
+        betas (tuple[float, float, float]): coefficient used for computing running averages of
+        update, square gradient and instability (default: (0.9, 0.999, 0.9999)))
+        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+    """
+
+    def __init__(
+        self,
+        params,
+        lr=None,
+        eps=(1e-30, 1e-16),
+        clip_threshold=1.0,
+        betas=(0.9, 0.999, 0.9999),
+        weight_decay=0.0,
+    ):
+        assert lr > 0.0
+        assert all([0.0 <= beta <= 1.0 for beta in betas])
+
+        defaults = dict(
+            lr=lr,
+            eps=eps,
+            clip_threshold=clip_threshold,
+            betas=betas,
+            weight_decay=weight_decay,
+        )
+        super(DistributedCAME, self).__init__(params, defaults)
+
+        self.distributed = False
+        self.zero = False
+        self.clip_method = dict()
+        self.ori_shape = dict()
+        self.working_shape = dict()
+        self.gather_before_compute = dict()
+        # record working parameter original shape (Before TP)
+        for group in self.param_groups:
+            for p in group["params"]:
+                self.ori_shape[id(p)] = p.data.shape
+
+    @property
+    def supports_memory_efficient_fp16(self):
+        return True
+
+    @property
+    def supports_flat_params(self):
+        return False
+
+    def setup_distributed(
+        self, master_to_working_map: dict, tp_group: ProcessGroup, zero_group: ProcessGroup, zero_flag: bool = False
+    ):
+        self.tensor_parallel_group = tp_group
+        self.zero_parallel_group = zero_group
+        # When running setup_distribute(), the parameters now are master parameters (After TP and zero)
+        # But master param != d_tensor, need to record information of working param (After TP before zero)
+        for group in self.param_groups:
+            for p in group["params"]:
+                # if no zero, working param == master param
+                working_param = master_to_working_map[id(p)] if master_to_working_map else p
+                self.ori_shape[id(p)] = self.ori_shape[id(working_param)]
+                self.working_shape[id(p)] = working_param.data.shape
+                self.gather_before_compute[id(p)] = False
+                try:
+                    api.get_device_mesh(working_param)
+                    sharding_spec = api.get_sharding_spec(working_param)
+                    self.clip_method[id(p)] = "col" if 0 in sharding_spec.dim_partition_dict.keys() else "row"
+                    # if dist.get_rank() == 0:
+                    #     print(self.ori_shape[id(p)], p.data.shape, sharding_spec.dim_partition_dict, self.clip_method[id(p)])
+                except:
+                    self.clip_method[id(p)] = None
+
+        self.zero = True if master_to_working_map and zero_flag else False
+        self.distributed = True
```

Comment:
What's the usage of this argument? It should be always True

Reply:
In the previous design, the DistributedCAME can also be used in non distribute environment, the difference is that if setup_distribued() is called, the optimizer will be labelled as distributed. And there are some communication operations different from CAME. Otherwise, the DistributedCAME is the same as CAME.